### ISSUE TITLE
A hook to access sharp instance for extra processing (crop, maintain EXIF, etc)

### ISSUE THREAD

**Issue by zachleat** [11ty/eleventy-img#52]
See https://twitter.com/ericvalois/status/1352203081304436738

**Comment by tusamni** [comment]
I'd love to be able to have a hook to access EXIF data.

**Comment by colingreen** [comment]
I'd like to see this too – I want to get hold of the image's dominant colour from the Sharp stats: https://sharp.pixelplumbing.com/api-input#stats

**Comment by zachleat** [comment]
> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy-img/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!

**Comment by ScreenDream** [comment]
I already commented on the other issue report, but want to repeat the request here:

Yes, keeping image metadata like EXIF, XMP, and IPTC and also the ICC colour profile are critical for me as well. 
I'm a visual artist building my website with eleventy ATM and will be unable to use the image plugin at all if I can't keep metadata and ICC profiles - my images look washed out and the contact information on the images is missing after the processing.

I would be super thankful if somebody could implement the hooks to activate/access the "withMetadata" functionality that is already in sharp: https://sharp.pixelplumbing.com/api-output#withmetadata
I am sadly not up to that myself.

Thanks a ton!

Tom

**Comment by philiprenich** [comment]
Would it be bad to do something as naïve as adding a `sharpHook` option that takes a Sharp instance as a parameter and returns a Sharp instance. Then in `img.js` (maybe around line 500?) have

```js
        if(typeof this.options.sharpHook === "function") {
          sharpInstance = this.options.sharpHook(sharpInstance);
        }
```

It's not foolproof by any means. It would allow the dev to completely override what 11ty-img is doing: one could `sharp.resize(100, null)` and it would ignore any resizing you've asked the plugin to do while not reporting correctly. Perhaps running the hook before everything else would be better. Still, I feel like there's some more user friendly way of doing this since this is pretty heavy-handed and permissive.

**Comment by ScreenDream** [comment]
I can't comment on the "would it be bad" part ;-)
But otherwise I think this could be interesting if nobody has a better and more "non-bad" solution while leaving the plugin intact otherwise.
Like a kind of "Pro-mode".

My current workaround is setting the default for "withMetadata" to "true" in the sharp/lib/constructor.js (line 261).
Dirty hack, but does what I need for the time being.
My website is almost ready now thanks to this.

But yeah, like I wrote on the other thread, IMO this is mission critical for any application of eleventy where the image IS the product.

Thanks to everybody looking into this!

**Comment by philiprenich** [comment]
@ScreenDream Feel free to use [my fork](https://github.com/philiprenich/eleventy-img/tree/naive-sharp-hook) - though I won't guarantee this will always be around, though for the moment I'll be using it until there's a solution in the main repo here (bit easier than re-writing my code to use Sharp directly).

Sounds like you're doing everything at build time though and don't need access to a proper package so your solution may be stable enough. 

[Here's how you'd use a GitHub url](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#github-urls) in your `package.json`

**Comment by hansspiess** [comment]
First of all, thanks for this great plugin - it is a really valuable tool when working with image content!

As this PR is closed: is the Feature (a hook exposing `sharpInstance`) being rejected? I think it would be a handy addition to the plugin, keeping it's API concise while opening possibilities for users.

**Comment by smth** [comment]
I was just wondering about this. I wrote up an [experiment here where I used Sharp to essentially get the background colour of an image](https://smth.uk/get-image-pixel-colours-in-eleventy-node/). I called Sharp directly to do this, which feels like a doubling of work (assuming the data I generated already exists somewhere).

**Comment by zachleat** [comment]
Shipping with v6.0.0-beta.1

````js
await Image("./test/exif-sample-large.jpg", {
  // Keep exif metadata
  transform: (sharp) => {
    sharp.keepExif();
  }
});
```

**Comment by ScreenDream** [comment]
Is 6.0.0-beta.1 available already?
I only see beta and release versions up to 5.0.0 with 
`npm show @11ty/eleventy-img versions`

And will that keep both Exif metadata and ICC profiles like "keepMetadata()" does with Sharp?

We are working on a new website and my old hack no longer works...

**Comment by zachleat** [comment]
Not yet—finishing up some last issue triage before we release! Soon!

**Comment by ScreenDream** [comment]
Awesome! Thank you! ❤️❤️❤️

**Comment by ScreenDream** [comment]
Ok, so I installed version 6.0.0-beta.2 today following your announcement on LinkedIn and while the code was accepted, I still did not have the full metadata in the output jpg, webp and avif files with the code you posted above, but when I replaced "keepExif()" with "keepMetadata()", it was all there.
AWESOME :-)
Thanks a ton!

Out of curiosity: Is this "transform: (sharp) => {...}" code a way to access everything sharp offers or is it only meant to be used for the metadata?

Cheers and a great 2025 to you!

Tom
